packages/firestore/src/api/user_data_writer.ts

```diff
@@ -17,99 +17,112 @@
+      return Number(value.doubleValue!);
+    } else if ('timestampValue' in value) {
+      return this.convertTimestamp(value.timestampValue!);
+    } else if ('stringValue' in value) {
+      return value.stringValue!;
+    } else if ('bytesValue' in value) {
+      return new Blob(normalizeByteString(value.bytesValue!));
+    } else if ('referenceValue' in value) {
+      return this.convertReference(value.referenceValue!);
+    } else if ('geoPointValue' in value) {
+      return new GeoPoint(
+        value.geoPointValue!.latitude!,
+        value.geoPointValue!.longitude!
+      );
+    } else if ('arrayValue' in value) {
+      return this.convertArray(value.arrayValue!);
+    } else if ('mapValue' in value) {
+      if (isServerTimestamp(value)) {
+        return this.convertServerTimestamp(value);
+      }
+      return this.convertObject(value.mapValue!);
     } else {
-      return value.value();
+      return fail('Invalid value type: ' + JSON.stringify(value));
     }
   }
 
-  private convertObject(data: ObjectValue): firestore.DocumentData {
+  private convertObject(mapValue: api.MapValue): firestore.DocumentData {
     const result: firestore.DocumentData = {};
-    data.forEach((key, value) => {
+    forEach(mapValue.fields || {}, (key, value) => {
       result[key] = this.convertValue(value);
     });
     return result;
   }
 
-  private convertArray(data: ArrayValue): unknown[] {
-    return data.internalValue.map(value => this.convertValue(value));
+  private convertArray(arrayValue: api.ArrayValue): unknown[] {
+    return (arrayValue.values || []).map(value => this.convertValue(value));
   }
 
-  private convertServerTimestamp(value: ServerTimestampValue): unknown {
+  private convertServerTimestamp(value: api.Value): unknown {
     switch (this.serverTimestampBehavior) {
       case 'previous':
-        return value.previousValue
-          ? this.convertValue(value.previousValue)
-          : null;
+        const previousValue = getPreviousValue(value);
+        if (previousValue == null) {
+          return null;
+        }
+        return this.convertValue(previousValue);
       case 'estimate':
-        return this.convertTimestamp(value.localWriteTime);
+        return this.convertTimestamp(getLocalWriteTime(value));
       default:
-        return value.value();
+        return null;
     }
   }
 
-  private convertTimestamp(timestamp: Timestamp): Timestamp | Date {
+  private convertTimestamp(value: api.Timestamp): Timestamp | Date {
+    const normalizedValue = normalizeTimestamp(value);
+    const timestamp = new Timestamp(
+      normalizedValue.seconds,
+      normalizedValue.nanos
+    );
     if (this.timestampsInSnapshots) {
       return timestamp;
     } else {
       return timestamp.toDate();
     }
   }
 
-  private convertReference(value: RefValue): DocumentReference<T> {
-    const key = value.value();
-    const database = this.firestore.ensureClientConfigured().databaseId();
```

Comment:
Losing this error message seems like a bad idea. Developers won't have any indication that this isn't going to work without this, no?

Reply:
I added this back. I was trying to limit the parsing code for references, since it is already pretty redundant in our code.